### DISCUSSION TITLE
Issues found by PVS-Studio

### DISCUSSION THREAD

**Discussion by skhrenov** [bulletphysics/bullet3#3542] [General]
Hello,
I have found a few bugs using PVS-Studio analyzer. PVS-Studio is a static code analyzer for C, C++ and C#: [https://www.viva64.com/en/pvs-studio/](https://www.viva64.com/en/pvs-studio/)

Analyzer warnings:

- [V773](https://www.viva64.com/en/w/V773/) The function was exited without releasing the 'memoryBuffer' pointer. A memory leak is possible. pybullet loadmeshfromstl.h 49
[A link to the source code.](https://github.com/bulletphysics/bullet3/blob/874d02d7aac4a562d3ef36d654e607bfbb372e68/examples/Importers/ImportSTLDemo/LoadMeshFromSTL.h#L49)

- [V773](https://www.viva64.com/en/w/V773/) The function was exited without releasing the 'tga_data' pointer. A memory leak is possible. pybullet stb_image.cpp 2969
[A link to the source code.](https://github.com/bulletphysics/bullet3/blob/874d02d7aac4a562d3ef36d654e607bfbb372e68/examples/ThirdPartyLibs/stb_image/stb_image.cpp#L2969)

- [V522](https://www.viva64.com/en/w/V522/) Dereferencing of the null pointer 'mbl' might take place. pybullet physicsservercommandprocessor.cpp 379
[A link to the source code.](https://github.com/bulletphysics/bullet3/blob/874d02d7aac4a562d3ef36d654e607bfbb372e68/examples/SharedMemory/PhysicsServerCommandProcessor.cpp#L379)

In addition, I suggest having a look at the emails, sent from @pvs-studio.com.

Best regards,
Sergey Hrenov,
PVS-Studio Team

**Comment by erwincoumans** [comment]
Thanks for the review, the 1st and 3rd are valid issues we need to fix. The 2nd seems pointless, since we check for 0 pointer, so there is nothing to be freed.

**Comment by erwincoumans** [comment]
Both issues are address in this commit:
https://github.com/bulletphysics/bullet3/commit/ad2099e107174759cf6d4d80469fd6c157858b89
Since the 2nd one isn't an issue, no action is needed.
Thanks!

**Comment by erwincoumans** [comment]
Thanks for the report! Most of the issues were in examples, which are less strictly developed.
Still, good to fix it, now it uses B3_PI:
https://github.com/bulletphysics/bullet3/blob/master/examples/SharedMemory/PhysicsClientC_API.cpp#L4145

Oh, and Bullet and PyBullet are still very actively developed and used, in particular for robotics and reinforcement learning.

**Comment by erwincoumans** [comment]
We should probably get rid of BussIK and also from the Bullet3/OpenCL code, it is not used.
Same for enet/tests etc.

**Comment by erwincoumans** [comment]
tinyrenderer example also fixed, for create cube x,y,z (and not x,y,y)

**Comment by erwincoumans** [comment]
I'll check it out. Until then, don't restart that example more than a trillion times :)

**Comment by erwincoumans** [comment]
Is there a way to exclude examples, in particular examples/ThirdPartyLibs? and exclude Bullet/src/Bullet3* ? The OpenCL pipeline is dead/unused, and I have no intention to fix ThirdPartyLibs.
